packages/timeline-state-resolver/src/integrations/tricaster/index.ts

```diff
@@ -0,0 +1,200 @@
+		this._doOnTime.on('slowSentCommand', (info) => this.emit('slowSentCommand', info))
+		this._doOnTime.on('slowFulfilledCommand', (info) => this.emit('slowFulfilledCommand', info))
+	}
+	async init(options: TriCasterOptions): Promise<boolean> {
+		const initPromise = new Promise<boolean>((resolve) => {
+			this._resolveInitPromise = resolve
+		})
+		this._connection = new TriCasterConnection(options.host, options.port ?? DEFAULT_PORT)
+		this._connection.on('connected', (info, shortcutStateXml) => {
+			this._stateDiffer = new TriCasterStateDiffer(info)
+			this._setInitialState(shortcutStateXml)
+			this._setConnected(true)
+			this._initialized = true
+			this._resolveInitPromise(true)
+			this.emit('info', `Connected to TriCaster ${info.productModel}, session: ${info.sessionName}`)
+		})
+		this._connection.on('disconnected', (_reason) => {
+			this._setConnected(false)
+		})
+		this._connection.on('error', (reason) => {
+			this.emit('error', 'TriCasterConnection', reason)
+		})
+		this._connection.connect()
+		return initPromise
+	}
+
+	private _setInitialState(shortcutStateXml: string): void {
+		if (!this._stateDiffer) {
+			throw new Error('State Differ not available')
+		}
+		const time = this.getCurrentTime()
+		const state = this._stateDiffer.externalStateConverter.getTriCasterStateFromShortcutState(shortcutStateXml)
+		this.setState(state, time)
+	}
+
+	private _connectionChanged(): void {
+		this.emit('connectionChanged', this.getStatus())
+	}
+
+	private _setConnected(connected: boolean): void {
+		if (this._connected !== connected) {
+			this._connected = connected
+			this._connectionChanged()
+		}
+	}
+
+	/** Called by the Conductor a bit before handleState is called */
+	prepareForHandleState(newStateTime: number): void {
+		// clear any queued commands later than this time:
+		this._doOnTime.clearQueueNowAndAfter(newStateTime)
+		this.cleanUpStates(0, newStateTime)
+	}
+
+	handleState(newState: TimelineState, newMappings: Mappings): void {
+		super.onHandleState(newState, newMappings)
+		if (!this._initialized || !this._stateDiffer) {
+			// before it's initialized don't do anything
+			this.emit('warning', 'TriCaster not initialized yet')
+			return
+		}
+
+		const previousStateTime = Math.max(this.getCurrentTime(), newState.time)
+		const oldState = this.getStateBefore(previousStateTime)?.state ?? this._stateDiffer.getDefaultState()
+
+		const newTriCasterState = this._stateDiffer.timelineStateConverter.getTriCasterStateFromTimelineState(
+			newState,
+			newMappings,
+			this.deviceId
+		)
+
+		const commandsToAchieveState = this._stateDiffer.getCommandsToAchieveState(newTriCasterState, oldState)
+
+		// clear any queued commands later than this time:
+		this._doOnTime.clearQueueNowAndAfter(previousStateTime)
+
+		// add the new commands to the queue:
+		this._addToQueue(commandsToAchieveState, newState.time)
+
+		// store the new state, for later use:
```

Comment:
This is the bit of logic that is (almost) identical in (almost) every integration, including the comments. I don't see a strong reason to remove them right now, as this will be gone anyway after we refactor to use the goodies from https://github.com/nrkno/sofie-timeline-state-resolver/pull/236